### PR TITLE
[CALCITE-2609]  Bugfix - adding additional requirement for JdbcFilter…

### PR DESCRIPTION
**Pull request by ptrbojko**
…Rule

Jdbc rules converts part of a query to a jdbc statement to external schema, but
when dynamic params are used it should either create prepared statement
with passed dynamic params or don't filter such cases. Simplier is to filter it.

After a bugfix one test in JdbcTest was obsolete because it relies on passing
the query to jdbc schema during planning. Test was trimmed a little bit. Part of it
was moved to another test.

This fixes also [CALCITE-2613] by adding simplest support for AvaticaParameter.getClassname().
This is done by implementing CalcitePrepareImpl.getClassName - simply by returning an Object
classname.

Fixing one of JdbcConverterRule constructors - it throws class cast exception. There was an
invalid casting from Guava Predicate (which is not a functional interface) to java buitin
Predicate (which is a functional interface), by choosing apply method as a functional interface
candidate.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java [hunk 1]
```diff
@@ -946,6 +957,26 @@ public boolean containsUserDefinedFunction() {
...
+    CheckingDynamicParamVisitor() {
+      super(true);
+    }
+
+    public boolean containsDynamicParam() {
+      return containsDynamicParam;
+    }
+
+    @Override public Void visitDynamicParam(RexDynamicParam dynamicParam) {
+      containsDynamicParam |= true;
```
Comment (eolivelli) [inline core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java]: Is this intentional? You are forcing the value to 'true'
Maybe a simple assignment is clearer

File: core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java [hunk 2]
```diff
@@ -943,7 +943,7 @@ private int getTypeOrdinal(RelDataType type) {
   }
 
   private static String getClassName(RelDataType type) {
-    return null;
+    return Object.class.getName();
```
Comment (eolivelli) [inline core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java]: Can we add a comment to explain this behavior?

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java [hunk 3]
```diff
@@ -6574,14 +6574,82 @@ public void testGetDate() throws Exception {
 
       rs.close();
 
-      final String sql = "update t2 set vals=? where id=?";
```
Comment (eolivelli) [inline core/src/test/java/org/apache/calcite/test/JdbcTest.java]: Why are you dropping this case?
I remember I recently (calcite 1.15) fixed a problem with jsbc params and UPDATE.

Can't we keep this test and add a new one

File: core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java (hunk 2 above)
Comment (ptrbojko) [reply core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java]: Will do

File: core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java (hunk 1 above)
Comment (ptrbojko) [reply core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java]: It is not intentional - I can simplify this.

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java (hunk 3 above)
Comment (ptrbojko) [reply core/src/test/java/org/apache/calcite/test/JdbcTest.java]: I am considering this a main issue of change. Before the patch former update query was planned in underlying jdbc schema as a subquery. After the patch - such query is not pushed to underying jdbc through jdbcrules logic, but stays in calcite. Calcite itself does not implement in a planner UDPATES - so it throws exception. 

I believe this part of a test was a little hack'y - it relied on the invalid behaviour of calcite planner.

Comment (eolivelli) [reply core/src/test/java/org/apache/calcite/test/JdbcTest.java]: From the change it seems to me that we are losing one of the most common cases, that is an UPDATE by PRIMARY KEY.
Maybe I am missing something.

Anyway it would be better to add an explicit testcase which tells what it is expected to happen with the query you dropped from the test case.
WDYT?

File: core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java [hunk 4]
```diff
@@ -943,7 +943,8 @@ private int getTypeOrdinal(RelDataType type) {
   }
 
   private static String getClassName(RelDataType type) {
-    return null;
+    // AvaticaParameter requires className field to be not null, see CALCITE-2613
+    return Object.class.getName();
```
Comment (joshelser) [inline core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java]: Is there a method on RelDataType which we could use already to determine the appropriate `Class`? If not, do you think it makes sense to introduce that to the `RelDataType` interface?

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java [hunk 5]
```diff
@@ -6574,14 +6574,82 @@ public void testGetDate() throws Exception {
...
+      calciteConnection.close();
+    }
+  }
+
+  /**
+   * Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-2609">[CALCITE-2609]
+   * Dynamic parameters ? pushed to underlying jdbc schema causing error </a>.
+   */
+  @Test public void testQueryWithParameter() throws Exception {
```
Comment (joshelser) [inline core/src/test/java/org/apache/calcite/test/JdbcTest.java]: Nice test :)

File: plus/pom.xml [hunk 6]
```diff
@@ -55,6 +55,14 @@ limitations under the License.
       <groupId>org.apache.calcite.avatica</groupId>
       <artifactId>avatica-core</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.apache.calcite.avatica</groupId>
+      <artifactId>avatica-server</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty</groupId>
+      <artifactId>jetty-server</artifactId>
```
Comment (joshelser) [inline plus/pom.xml]: I'm not seeing any explicit additions of jetty-server classes to `plus/src` as a part of this changeset. Why did you add `jetty-server` here?

File: plus/src/main/java/org/apache/calcite/chinook/ChinookAvaticaServer.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.List;
+
+/**
+ * Wrapping Calcite engine with Avatica tansport for testing JDBC capabilities between
+ * Avatica JDBC transport and Calcite
+ */
+public class ChinookAvaticaServer {
```
Comment (joshelser) [inline plus/src/main/java/org/apache/calcite/chinook/ChinookAvaticaServer.java]: Nice!

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java (hunk 3 above)
Comment (ptrbojko) [reply core/src/test/java/org/apache/calcite/test/JdbcTest.java]: I will try to look at this next week. Maybe rewrite the mentioned test not to use the Connection.prepareStatement but some internals.

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java (hunk 5 above)
Comment (ptrbojko) [reply core/src/test/java/org/apache/calcite/test/JdbcTest.java]: Bug reproduction :)

File: core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java (hunk 4 above)
Comment (ptrbojko) [reply core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java]: There is no mapping from SQL types to Java types now. Having it (maybe on RelDataType level, maybe simple static mapping somewhere close to SqlTypeName) could be a good thing but I think that Calcite use cases for that narrows only to satisfy exposing through avatica.  Although this propably will change and then some can think how to implement the mapping.

File: plus/pom.xml (hunk 6 above)
Comment (ptrbojko) [reply plus/pom.xml]: This is for satisfying the maven checks. Avatica server indirectly imports this, so the checker fails with indirect dependency.

Comment (joshelser) [reply plus/pom.xml]: The point of transitive dependencies is that they come in automatically. You should not add them as a dependency unless you are using classes from them. What is the error you're seeing without this?

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java (hunk 3 above)
Comment (ptrbojko) [reply core/src/test/java/org/apache/calcite/test/JdbcTest.java]: @eolivelli  I've recreated the test case. But now rather based on PreparedStatement it uses directly planner and traversing through Rel nodes. 

The former test worked by hack - it was relying on that the whole query was pushed to jdbc schema, thus omitting the ling4j implementation. 

When PreparedStatement is being prepared - whole query is compiled and update clause is not supported. So the former test worked only because update was pushed to jdbc subschema. 

Current tests - checks the same query as former and validates whether dynamic parameters exists and are specified. But it uses for that traversing through rel nodes directly.

I hope this clarifies the thing.

**Comment by ptrbojko** [comment]
I recall this pull request, see CALCITE-2609 and CALCITE-563
